src/ray/object_manager/object_manager.h

```diff
@@ -0,0 +1,278 @@
+
+  /// Subscribe to notifications of objects deleted from local store.
+  /// \param callback The callback to invoke when objects are removed from the local store.
+  /// \return Status of whether adding the subscription succeeded.
+  ray::Status SubscribeObjDeleted(std::function<void(const ray::ObjectID&)> callback);
+
+  /// Push an object to to the node manager on the node corresponding to client id.
+  /// \param object_id The object's object id.
+  /// \param client_id The remote node's client id.
+  /// \return Status of whether the push request successfully initiated.
+  ray::Status Push(const ObjectID &object_id,
+                   const ClientID &client_id);
+
+  /// Pull an object from ClientID. Returns UniqueID asociated with
+  /// an invocation of this method.
+  /// \param object_id The object's object id.
+  /// \return Status of whether the pull request successfully initiated.
+  ray::Status Pull(const ObjectID &object_id);
+
+  /// Discover ClientID via ObjectDirectory, then pull object
+  /// from ClientID associated with ObjectID.
+  /// \param object_id The object's object id.
+  /// \param client_id The remote node's client id.
+  /// \return Status of whether the pull request successfully initiated.
+  ray::Status Pull(const ObjectID &object_id,
+                   const ClientID &client_id);
+
+  /// Add a connection to a remote object manager.
+  /// This is invoked by an external server.
+  /// \param conn The connection.
+  /// \return Status of whether the connection was successfully established.
+  ray::Status AcceptConnection(TCPClientConnection::pointer conn);
+
+  /// Cancels all requests (Push/Pull) associated with the given ObjectID.
+  /// \param object_id The ObjectID.
+  /// \return Status of whether requests were successfully cancelled.
+  ray::Status Cancel(const ObjectID &object_id);
+
+  // Callback definition for wait.
+  using WaitCallback = std::function<void(const ray::Status,
+                                          uint64_t,
+                                          const std::vector<ray::ObjectID>&)>;
+  /// Wait for timeout_ms before invoking the provided callback.
+  /// If num_ready_objects is satisfied before the timeout, then
+  /// invoke the callback.
+  /// \param object_ids The object ids to wait on.
+  /// \param timeout_ms The time in milliseconds to wait before invoking the callback.
+  /// \param num_ready_objects The minimum number of objects required before
+  /// invoking the callback.
+  /// \param callback Invoked when either timeout_ms is satisfied OR num_ready_objects
+  /// is satisfied.
+  /// \return Status of whether the wait successfully initiated.
+  ray::Status Wait(const std::vector<ObjectID> &object_ids,
+                   uint64_t timeout_ms,
+                   int num_ready_objects,
+                   const WaitCallback &callback);
+
+  /// \return Whether this object was successfully terminated.
+  ray::Status Terminate();
+
+ private:
```

Comment:
(1) Do these have to be private methods?  If they are just helpers, put them in the cc file, and surround with `namespace { }`.
(2) The reason being the *user* of this class will glance at the header file, and the user should not be bothered with implementation details.